sdk/servicebus/service-bus/src/sender.ts

```diff
@@ -88,7 +88,11 @@ export class Sender {
    * want to front load the work of setting up the AMQP link to the service.
    */
   async open(): Promise<void> {
-    return MessageSender.create(this._context).open();
+    await MessageSender.create(this._context).open();
+
+    if (this._context.sender?.isOpen) {
+      this._isClosed = false;
+    }
```

Comment:
Should we pull this up before calling `await MessageSender.create(this._context).open();`?
What if send() gets called before `open()` is resolved? They would get the "sender is closed error"

Reply:
Isn't that accurate though - send() has no guarantee that open() is complete and it wouldn't be accurate to say it's open _before_ the open call? 

As an aside, `MessageSender.open()` does a similar check internally so we do actually have a check before and after (although it's through an async call, so not necessarily as optimized as it could be).

I'll wait to merge until your reply, just in case I'm missing some nuance.

Reply:
> send() has no guarantee that open() is complete and it wouldn't be accurate to say it's open before the open call

send() would call the internal open and wait to get hold of a lock as long as `_isClosed` is not true. If the user has called `open()`, then I'd rather have the following `send()` call wait to get hold of the lock than throw the "sender is closed" error

> As an aside, MessageSender.open() does a similar check internally so we do actually have a check before and after (although it's through an async call, so not necessarily as optimized as it could be).

How are they the same? The `isOpen()` check inside `MessageSender.open()` is done to decide whether the underlying AMQP link should be opened or not. On the outside `Sender.send()` method, the check being made is whether the user called `close()` on the sender/client or not

Using code to visualize, I am talking about the below case
```
sender.close();
sender.open();
sender.send();
```

My concern is that in the above case, the user will get `The sender for "${entityPath}" has been closed and can no longer be used. Please create a new sender using the "getSender" method on the ServiceBusClient.` Whereas, ideally we want the send() to wait for the link to be opened up and then use it

Reply:
That seems counter to how I would expect this to work.

`open()` explicitly returns a Promise to be await'ed on to validate that the open succeeded - why would we not use that or expect the user to use that? If the user calls other methods while the object is in an indeterminate state throwing seems entirely appropriate.

Reply:
`sender.open()` has started the process of opening up the sender. 

The error message when calling `sender.send()` recommends that the user should create a new sender instead. If we do indeed want an error to be thrown, it should recommend that the sender is in the process of opening and that the user should retry after a while.

On the other hand, if we don't throw the error, then the `send()` will wait to get hold of the lock, which it should be able to get right after `open()` completes and the send the message successfully.

Reply:
Let me look at that error message - it doesn't quite make sense, does it? :)

Reply:
Updated - there was a branch in the code for whether it was the connection that was closed vs the individual sender. I _only_ updated the individual sender branch because I don't think re-opening a Sender should (in any way) reconstitute the entire connection.

Fix: [bf26f0](https://github.com/Azure/azure-sdk-for-js/pull/8467/commits/bf26f0e6d5e8e29cd963f0f1191a87d708a8f50a)

Reply:
After looking at this more with @ramya-rao-a we've decided to go the other direction - close()ing closes and open() does _not_ reopen it.

The issue is that there are some complications in the design that arise from being able to re-open a sender when you start closing at different areas of the client hierarchy. 
- Do we want to be able to undo the .close() at a QueueClient/TopicClient level from the sender?
- Same for ServiceBusClient - if I close the parent should Sender.open() undo it?

So for now the simplest solution is to maintain the design we have. close() is a final operation.